### PR TITLE
Introducing public header directory for OrbitCore, OrbitGl and OrbitAsm

### PR DESCRIPTION
**Pull request by beckerhe**
Currently all the modules are exposing their top-level directory as its external include directory.

This PR is going to change that by introducing a dedicated include path subdirectory `include/`.
All public headers (for the sake of simplicity all the headers for now) were moved into this
subdirectory. The directory structure now looks like this:

```bash
<module>/
<module>/impl1.cpp
<module>/impl2.cpp
<module>/include/
<module>/include/<module>/
<module>/include/<module>/header1.h
<module>/include/<module>/header2.h
```

When I want to include `header1.h` from `<module>` in a different module, I write the following:
```C
#include "<module>/header.1h"
```

It tried compiling on Windows and Linux. Both builds should work and benefit from this change.

These changes uncovered some other problems, which are resolved in here in part:
- `gte` didn't have a `Findgte.cmake` file and was included by direct path.
- `Orbit.h` is not included from anywhere, so I removed the file.
- `bin/msdia140.dll` is not used and probably leads to license issues, since it's a proprietary library. So I removed it.

### CODE REVIEW

File: OrbitCore/CMakeLists.txt [hunk 1]
```diff
@@ -148,17 +151,17 @@ if(WIN32)
...
+           ${HEADERS}/BpfTrace.h
+           ${HEADERS}/LibunwindstackUnwinder.h
+           ${HEADERS}/LinuxEventTracer.h
+           ${HEADERS}/LinuxPerfEvent.h
+           ${HEADERS}/LinuxPerfEventProcessor.h
+           ${HEADERS}/LinuxPerfEventVisitor.h
+           ${HEADERS}/LinuxPerfRingBuffer.h
+           ${HEADERS}/LinuxPerfUtils.h
+           ${HEADERS}/LinuxUprobesUnwindingVisitor.h
+           ${HEADERS}/LinuxUtils.h)
```
Comment (dimitry-) [inline OrbitCore/CMakeLists.txt]: Maybe it is better not to hide these behind a variable... Ideally replace with something like everything under include/ dir? does cmake support this?

File: OrbitCore/Hijacking.cpp [hunk 2]
```diff
@@ -13,7 +13,7 @@
 #include "Core.h"
 #include "Message.h"
 #include "MinHook.h"
-#include "OrbitAsm.h"
+#include "OrbitAsm/OrbitAsm.h"
```
Comment (dimitry-) [inline OrbitCore/Hijacking.cpp]: can you merge this commit into the one that broke windows build?

**Comment by pierricgimmig** [comment]
Hi Henning, I'm happy with specifying the module when including a header:

> ```c
> #include "<module>/header1.h"
> ```

However, I'm really not a huge fan of this pattern: *<module>/include/<module>/header1.h*.  Would keeping the current directory structure and putting whatever header we'd want to hide in an "internal" or "private" sub-directory achieve the same result?  We'd leave the file structure unchanged, the public headers would be in the <module> directory, we'd keep things simple...  

> Currently all the modules are exposing their top-level directory as its external include directory.
What is the problem here?  Are there occurrences in the codebase where a module includes headers from another module that it shouldn't be including?  Is this to facilitate an eventual release of individual libraries?

Concerning msdia140.dll, removing it essentially makes the Windows version unusable.  On Windows it is used to parse the pdb files.  Unless/until we have an alternative, we can't simply remove it.

Orbit.h can be used as a header-only library to enable manual instrumentation on Windows.  It's not meant to be used by Orbit itself.  We could potentially move it to another directory, OrbitAPI or OrbitSDK maybe, but I wouldn't remove it.

**Comment by beckerhe** [comment]
> Hi Henning, I'm happy with specifying the module when including a header:
> 
> > ```c
> > #include "<module>/header1.h"
> > ```
> 
> However, I'm really not a huge fan of this pattern: _module_/include/_module_/header1.h*. Would keeping the current directory structure and putting whatever header we'd want to hide in an "internal" or "private" sub-directory achieve the same result? We'd leave the file structure unchanged, the public headers would be in the _module_ directory, we'd keep things simple...

Yeah, I know the _module_/include/_module_/header-thing is ugly. But it is the only pattern that works, at least the only one I'm aware of.

I don't think having an "internal" subdirectory solves the problem. But first things first:

 * If we wanna have `#include "_module_/header.h"`-includes, there will be two options. Either make the top-level directory of the repo an include directory or the pattern I suggest. Going with the former approach, I believe it will make the situation worse, because you can't differentiate anymore between modules. You can always include a header file from any module in the project, even if your current file does not link against this module. Your IDE will suggest you to include a file, which in the end results in a linking error. I think that's makes it more error prone and discouraging for everybody not familiar with the codebase.
* The only alternative to the `#include "_module_/header.h"` is to stick to the `#include "header.h"`, which we have right now. This makes it less explicit (you never know, where a header file comes from) and it is likely to have name clashes and weird bugs (I've seen this with header names like `utils.h` or `log.h` which might appear in a third party dependency and in the project's codebase itself.)

> 
> > Currently all the modules are exposing their top-level directory as its external include directory.
> 
> What is the problem here? Are there occurrences in the codebase where a module includes headers from another module that it shouldn't be including? Is this to facilitate an eventual release of individual libraries?

As far as I'm aware, there is no intention to package individual libraries. My intention with this is
to help untangle the modules a bit more. I wanted to remove some cyclic dependencies and ended
up suggesting this first. Regarding your question, if there are modules which include header files,
they shouldn't include: Yes, there are some. There have been several cyclic dependencies in the code base and I was able to clean up some of them, while making these header change. To name a few:

*  OrbitPlugin/OrbitUserData.h needs things from OrbitCore and OrbitCore needs OrbitUserData.h, hence a cyclic dependency. So I had to move OrbitUserData.h, OrbitData.h and OrbitSDK.h to OrbitCore.
* minhook depends on OrbitAsm. And OrbitCore needs minhook, but OrbitAsm also needs OrbitCore. This is not solved yet.
* OrbitGl depends on OrbitAsm, but OrbitAsm is windows only, so it shouldn't include its headers. That's for now solved by a hack in cmake.
* There was more involving OrbitDll, but I can't remember right now. I'm sorry.

Apart from that, I believe that having a header not public by default helps thinking about the API
of a module. This leads to better structured code and better code in general. Additionally it will help
us refactorting OrbitGl, such that we will eventually have a OrbitService which does not depend on
OpenGL anymore.

> 
> Concerning msdia140.dll, removing it essentially makes the Windows version unusable. On Windows it is used to parse the pdb files. Unless/until we have an alternative, we can't simply remove it.

If I'm not mistaken it comes from the bin/ folder in the DIA SDK. Can't we just take it from there?
It could add a cmake rule, copying over this file. And is the dll not tied to a specific version of the
DIA SDK? I just checked the DLLs from 2017 Professional and 2019 Professional and they are
different (granted I just checked the file size.)
> 
> Orbit.h can be used as a header-only library to enable manual instrumentation on Windows. It's not meant to be used by Orbit itself. We could potentially move it to another directory, OrbitAPI or OrbitSDK maybe, but I wouldn't remove it.

Alright, got it. Will move it to a different directory. So this is something what we will eventually ship as part of a binary package of Orbit, right? Such that the user can include it in its build? I could add a cmake target which installs the header as a header-only library and can have cmake generate proper cmake file (such that you can include it in a cmake-based project with just a single find_package call.) What do you think?

**Comment by pierricgimmig** [comment]
TL;DR: I hate it, but OK ;-)

> Yeah, I know the _module_/include/_module_/header-thing is ugly. But it is the only pattern that works, at least the only one I'm aware of.

Anyone can still include "../../whatever_i_want.h" right :-) ?  If it were bulletproof, then maybe, but to be honest, I feel like this is overkill. The module/include/module/header-thing hurts my eyes, I really don't like it.

> Either make the top-level directory of the repo an include directory or the pattern I suggest. Going with the former approach, I believe it will make the situation worse, because you can't differentiate anymore between modules.

I think that former approach is perfectly fine actually.  From low to high level (and in alphabetical order): OrbitAsm, OrbitCore, OrbitGl, OrbitQt.  The rule is: don't include from a higher level.  I think we can easily enforce it in code reviews, just like we'd enforce not using relative paths.  In most cases the build would fail with linking errors.  We can still use the "module/header" pattern here.  

@beckerhe , you know I really appreciate your contributions and the thought you put in every PR.  Regarding this particular change, between our two diverging opinions I'm really tempted to choose my own (keeping the headers next to the cpp files).  But... I'm curious to see if your solution will indeed make the code cleaner by having public and private headers, so let's go for it.

> If I'm not mistaken it comes from the bin/ folder in the DIA SDK. Can't we just take it from there?
> It could add a cmake rule, copying over this file. And is the dll not tied to a specific version of the
> DIA SDK?

This is tricky, there are many versions of the msdia dll varying hugely in performance.  It is important to use this exact one for now.  This could potentially be replaced by llvm actually if there is feature parity for parsing pdbs.

> So this is something what we will eventually ship as part of a binary package of Orbit, right? Such that the user can include it in its build? I could add a cmake target which installs the header as a header-only library and can have cmake generate proper cmake file (such that you can include it in a cmake-based project with just a single find_package call.) What do you think?

Hmmm, that sounds like a lot of work for a single header!!  I wouldn't worry about it too much for now.  Let's wait until we re-think how we handle manual instrumentation.

**Comment by beckerhe** [comment]
Hi Pierric,
there is no need to merge something you hate.

When I wrote these changes, I was totally aware that
this might not get through. And it didn't involve a lot of work.
I would have discussed this change before hand, but you know,
time shifts. So I decided to make a proposal in the form of a PR.

I'm sorry for the headaches. Going to close this! :-)
